2.4.9 generates bad output for 'An [issue1](issue1) link'

Using python 3.6, 3.8, 3.10, 3.11 and 3.12 with 2.4.9 the string:

```
'An [issue1](issue1) link'
```
generates:
```
'<p>An <a href="#">issue1</a> link</p>\n'
```
when I expect:
```
'<p>An <a  rel="nofollow" href="issue1">issue1</a> link</p>\n'
```

This works as expected with 2.4.8 with my local changes.

I am subclassing Markdown2 to disable hyperlinks for some schemes:
```
_disable_url_schemes = ['javascript', 'data']
```
The code reads (from https://github.com/roundup-tracker/roundup/blob/master/roundup/cgi/templating.py#L59-L81):
```
def _import_markdown2():
    try:
        import markdown2
        import re

        class Markdown(markdown2.Markdown):
            # don't allow disabled protocols in links
            _safe_protocols = re.compile('(?!' + ':|'.join([
                re.escape(s) for s in _disable_url_schemes])
                                         + ':)', re.IGNORECASE)

        def _extras(config):
            extras = {'fenced-code-blocks': {}, 'nofollow': None}
            if config['MARKDOWN_BREAK_ON_NEWLINE']:
                extras['break-on-newline'] = True
            return extras

        markdown = lambda s, c: Markdown(safe_mode='escape', extras=_extras(c)).convert(s)  # noqa: E731
    except ImportError:
        markdown = None

    return markdown
```
When it is called in this test case, MARKDOWN_BREAK_ON is False and `_extras(c)` results in this call:
```
Markdown(safe_mode='escape', extras={'fenced-code-blocks': {}, 'nofollow': None}).convert('An [issue1](issue1) link')
```

I was able to reproducible it even without my shenanigans with this simple test:
```
>>> import markdown2
>>> markdown2.Markdown(safe_mode='escape', extras={'fenced-code-blocks': {}, 'nofollow': None}).convert('An [issue1](issue1) link')
'<p>An <a href="#">issue1</a> link</p>\n'
```
failure. This failed as well with 2.4.8 and only passed by going back to 2.3.3.

I saw couple of changes to link processing in the change notes:
* [pull #513] Fix relative links not working in safe mode (issue #254)

seems the most likely to cause this and also identified and early 2.3 release as working correctly.

@nicholasserra, the original failing test passed, but this test is now failing 8-):
```
        p = StringHTMLProperty(self.client, 'test', '1', None, 'test', u2s(u'An [issue1](https://example.com/issue1) link'))
>       self.assertIn( u2s(u'href="https://example.com/issue1"'), p.markdown().strip())
E       AssertionError: 'href="https://example.com/issue1"' not found in '<p>An <a href="#">issue1</a> link</p>'
```
the markdown input (p) is: `'An [issue1](https://example.com/issue1) link'`
markdown2 HTML output is: `'<p>An <a href="#">issue1</a> link</p>\n'`
so it looks like we just moved the error to the fully specified version 8-(.

I used:
```
python3 -m pip install -U  git+https://github.com/Crozzers/python-markdown2.git@fix-issue-517
```
to install the update. Running `python3 -c 'import markdown2; print(markdown2.__version__)'` returns `2.4.10`. So it looks like the install update was successful. I started with 2.4.8, pip updated to 2.4.9 an then to 2.4.10. The tests passed, failed, failed.

-- rouilj

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -2,7 +2,7 @@
 
 ## python-markdown2 2.4.10 (not yet released)
 
-(nothing yet)
+- [pull #520] Allow more relative links in safe mode (issue #517)
 
 
 ## python-markdown2 2.4.9

--- a/lib/markdown2.py
+++ b/lib/markdown2.py
@@ -1507,7 +1507,7 @@ class Markdown(object):
         self._escape_table[url] = key
         return key
 
-    _safe_protocols = r'(?:https?|ftp):\/\/|(?:mailto|tel):|\/|\.{,2}|#'
+    _safe_protocols = r'(?:https?|ftp):\/\/|(?:mailto|tel):'
 
     @property
     def _safe_href(self):
@@ -1518,14 +1518,14 @@ class Markdown(object):
         Modifications and bugfixes (c) 2009 Dana Robinson
         Modifications and bugfixes (c) 2009-2014 Stack Exchange Inc.
         '''
-        safe = r'-\w/'
+        safe = r'-\w'
         # omitted ['"<>] for XSS reasons
-        less_safe = r'\.!#$%&\(\)\+,/:;=\?@\[\]^`\{\}\|~'
+        less_safe = r'#/\.!#$%&\(\)\+,/:;=\?@\[\]^`\{\}\|~'
         # dot seperated hostname, optional port number, not followed by protocol seperator
         domain = r'(?:[%s]+(?:\.[%s]+)*)(?::\d+/?)?(?![^:/]*:/*)' % (safe, safe)
         fragment = r'[%s]*' % (safe + less_safe)
 
-        return re.compile(r'^(%s)?(%s)(%s)$' % (self._safe_protocols, domain, fragment), re.I)
+        return re.compile(r'^(?:(%s)?(%s)(%s)|(#|\.{,2}/)(%s))$' % (self._safe_protocols, domain, fragment, fragment), re.I)
 
     def _do_links(self, text):
         """Turn Markdown link shortcuts into XHTML <a> and <img> tags.

--- a/lib/markdown2.py
+++ b/lib/markdown2.py
@@ -1509,9 +1509,8 @@ class Markdown(object):
 
     _safe_protocols = r'(?:https?|ftp):\/\/|(?:mailto|tel):|\/|\.{,2}|#'
 
-    @classmethod
     @property
-    def _safe_href(cls):
+    def _safe_href(self):
         '''
         _safe_href is adapted from pagedown's Markdown.Sanitizer.js
         From: https://github.com/StackExchange/pagedown/blob/master/LICENSE.txt
@@ -1526,7 +1525,7 @@ class Markdown(object):
         domain = r'(?:[%s]+(?:\.[%s]+)*)(?::\d+/?)?(?![^:/]*:/*)' % (safe, safe)
         fragment = r'[%s]*' % (safe + less_safe)
 
-        return re.compile(r'^(%s)?(%s)(%s)$' % (cls._safe_protocols, domain, fragment), re.I)
+        return re.compile(r'^(%s)?(%s)(%s)$' % (self._safe_protocols, domain, fragment), re.I)
 
     def _do_links(self, text):
         """Turn Markdown link shortcuts into XHTML <a> and <img> tags.

--- a/lib/markdown2.py
+++ b/lib/markdown2.py
@@ -1509,8 +1509,9 @@ class Markdown(object):
 
     _safe_protocols = r'(?:https?|ftp):\/\/|(?:mailto|tel):|\/|\.{,2}|#'
 
+    @classmethod
     @property
-    def _safe_href(self):
+    def _safe_href(cls):
         '''
         _safe_href is adapted from pagedown's Markdown.Sanitizer.js
         From: https://github.com/StackExchange/pagedown/blob/master/LICENSE.txt
@@ -1525,7 +1526,7 @@ class Markdown(object):
         domain = r'(?:[%s]+(?:\.[%s]+)*)(?::\d+/?)?(?![^:/]*:/*)' % (safe, safe)
         fragment = r'[%s]*' % (safe + less_safe)
 
-        return re.compile(r'^(%s)?(%s)(%s)$' % (self._safe_protocols, domain, fragment), re.I)
+        return re.compile(r'^(%s)?(%s)(%s)$' % (cls._safe_protocols, domain, fragment), re.I)
 
     def _do_links(self, text):
         """Turn Markdown link shortcuts into XHTML <a> and <img> tags.

--- a/lib/markdown2.py
+++ b/lib/markdown2.py
@@ -1522,7 +1522,7 @@ class Markdown(object):
         # omitted ['"<>] for XSS reasons
         less_safe = r'#/\.!#$%&\(\)\+,/:;=\?@\[\]^`\{\}\|~'
         # dot seperated hostname, optional port number, not followed by protocol seperator
-        domain = r'(?:[%s]+(?:\.[%s]+)*)(?::\d+/?)?(?![^:/]*:/*)' % (safe, safe)
+        domain = r'(?:[%s]+(?:\.[%s]+)*)(?:(?<!tel):\d+/?)?(?![^:/]*:/*)' % (safe, safe)
         fragment = r'[%s]*' % (safe + less_safe)
 
         return re.compile(r'^(?:(%s)?(%s)(%s)|(#|\.{,2}/)(%s))$' % (self._safe_protocols, domain, fragment, fragment), re.I)

--- a/lib/markdown2.py
+++ b/lib/markdown2.py
@@ -1507,12 +1507,25 @@ class Markdown(object):
         self._escape_table[url] = key
         return key
 
-    # _safe_href is copied from pagedown's Markdown.Sanitizer.js
-    # From: https://github.com/StackExchange/pagedown/blob/master/LICENSE.txt
-    # Original Showdown code copyright (c) 2007 John Fraser
-    # Modifications and bugfixes (c) 2009 Dana Robinson
-    # Modifications and bugfixes (c) 2009-2014 Stack Exchange Inc.
-    _safe_href = re.compile(r'^((https?|ftp):\/\/|\/|\.|#)[-A-Za-z0-9+&@#\/%?=~_|!:,.;\(\)*[\]$]*$', re.I)
+    _safe_protocols = r'(?:https?|ftp):\/\/|(?:mailto|tel):|\/|\.{,2}|#'
+
+    @property
+    def _safe_href(self):
+        '''
+        _safe_href is adapted from pagedown's Markdown.Sanitizer.js
+        From: https://github.com/StackExchange/pagedown/blob/master/LICENSE.txt
+        Original Showdown code copyright (c) 2007 John Fraser
+        Modifications and bugfixes (c) 2009 Dana Robinson
+        Modifications and bugfixes (c) 2009-2014 Stack Exchange Inc.
+        '''
+        safe = r'-\w/'
+        # omitted ['"<>] for XSS reasons
+        less_safe = r'\.!#$%&\(\)\+,/:;=\?@\[\]^`\{\}\|~'
+        # dot seperated hostname, optional port number, not followed by protocol seperator
+        domain = r'(?:[%s]+(?:\.[%s]+)*)(?::\d+/?)?(?![^:/]*:/*)' % (safe, safe)
+        fragment = r'[%s]*' % (safe + less_safe)
+
+        return re.compile(r'^(%s)?(%s)(%s)$' % (self._safe_protocols, domain, fragment), re.I)
 
     def _do_links(self, text):
         """Turn Markdown link shortcuts into XHTML <a> and <img> tags.

--- a/test/tm-cases/link_safe_urls.html
+++ b/test/tm-cases/link_safe_urls.html
@@ -1,11 +1,47 @@
+<h1>Normal links</h1>
+
 <p><a href="https://www.example.com">Safe link 1</a></p>
 
 <p><a href="http://www.example.com">Safe link 2</a></p>
 
 <p><a href="ftp://www.example.com">Safe link 3</a></p>
 
-<p><a href="#anchor">Safe link 4</a></p>
+<p><a href="mailto:emailaddress@server.com">Safe link 4</a></p>
+
+<p><a href="tel:0123456789">Safe link 5</a></p>
+
+<p><a href="#anchor">Safe link 6</a></p>
+
+<h1>Bad protocols</h1>
 
 <p><a href="#">Unsafe link 1</a></p>
 
 <p><a href="#">Unsafe link 2</a></p>
+
+<h1>Relative links</h1>
+
+<p><a href="example">Safe link 1</a></p>
+
+<p><a href="./example">Safe link 2</a></p>
+
+<p><a href="../../example">Safe link 3</a></p>
+
+<p><a href="/example">Safe link 4</a></p>
+
+<p><a href="#">Unsafe link 1</a></p>
+
+<h1>Edge cases</h1>
+
+<p><a href="www.example.com/abc:def">Safe link 1</a></p>
+
+<p><a href="www.example.com/abc:/def">Safe link 2</a></p>
+
+<p><a href="https://www.example.com:4200">Safe link 3</a></p>
+
+<p><a href="https://www.example.com:4200/abcdef">Safe link 4</a></p>
+
+<p><a href="#">Unsafe link 1</a></p>
+
+<p><a href="#">Unsafe link 2</a></p>
+
+<p><a href="#">Unsafe link 3</a></p>

--- a/test/tm-cases/link_safe_urls.text
+++ b/test/tm-cases/link_safe_urls.text
@@ -1,11 +1,47 @@
+# Normal links
+
 [Safe link 1](https://www.example.com)
 
 [Safe link 2](http://www.example.com)
 
 [Safe link 3](ftp://www.example.com)
 
-[Safe link 4](#anchor)
+[Safe link 4](mailto:emailaddress@server.com)
+
+[Safe link 5](tel:0123456789)
+
+[Safe link 6](#anchor)
+
+# Bad protocols
 
 [Unsafe link 1](unknown://www.example.com)
 
-[Unsafe link 2](example)
+[Unsafe link 2](mailfrom:www.example.com)
+
+# Relative links
+
+[Safe link 1](example)
+
+[Safe link 2](./example)
+
+[Safe link 3](../../example)
+
+[Safe link 4](/example)
+
+[Unsafe link 1](.../www.example.com)
+
+# Edge cases
+
+[Safe link 1](www.example.com/abc:def)
+
+[Safe link 2](www.example.com/abc:/def)
+
+[Safe link 3](https://www.example.com:4200)
+
+[Safe link 4](https://www.example.com:4200/abcdef)
+
+[Unsafe link 1](unknown://www.example.com://abc)
+
+[Unsafe link 2](C:/Windows/System32)
+
+[Unsafe link 3](C:\Windows\System32)

